File: product_research_app/static/js/winner_v2.js
(function(global){
  const metricDefs = [
    {key:'magnitud_deseo',label:'Magnitud deseo',tip:'Nivel de deseo del producto'},
    {key:'nivel_consciencia_headroom',label:'Headroom de consciencia',tip:'Conocimiento del problema/solución'},
    {key:'evidencia_demanda',label:'Evidencia demanda',tip:'Demanda observada'},
    {key:'tasa_conversion',label:'Tasa conversión',tip:'Porcentaje de conversión'},
    {key:'ventas_por_dia',label:'Ventas por día',tip:'Unidades vendidas por día'},
    {key:'recencia_lanzamiento',label:'Recencia lanzamiento',tip:'Tiempo desde lanzamiento'},
    {key:'competition_level_invertido',label:'Competencia (invertido)',tip:'Nivel de competencia inverso'},
    {key:'facilidad_anuncio',label:'Facilidad anuncio',tip:'Facilidad para anunciar'},
    {key:'escalabilidad',label:'Escalabilidad',tip:'Capacidad de escalar'},
    {key:'durabilidad_recurrencia',label:'Durabilidad/recurrencia',tip:'Durabilidad o recurrencia'}
  ];

  const MAPS = {
    magnitud_deseo:{low:0.33, medium:0.66, high:1.0},
    nivel_consciencia_headroom:{unaware:1, problem:0.8, solution:0.6, product:0.4, most:0.2},
    competition_level_invertido:{low:1.0, medium:0.5, high:0.0},
    facilidad_anuncio:{low:0.33, med:0.66, medium:0.66, high:1.0},
    escalabilidad:{low:0.33, med:0.66, medium:0.66, high:1.0},
    durabilidad_recurrencia:{consumible:1.0, durable:0.0, intermedio:0.5}
  };

  function getDefaultWeightsV2(){
    return {
      lanzamiento:{
        magnitud_deseo:70,
        nivel_consciencia_headroom:60,
        evidencia_demanda:50,
        tasa_conversion:40,
        ventas_por_dia:50,
        recencia_lanzamiento:70,
        competition_level_invertido:60,
        facilidad_anuncio:60,
        escalabilidad:40,
        durabilidad_recurrencia:40
      },
      rentabilidad:{
        evidencia_demanda:80,
        ventas_por_dia:70,
        tasa_conversion:60,
        durabilidad_recurrencia:60,
        competition_level_invertido:50,
        magnitud_deseo:50,
        nivel_consciencia_headroom:40,
        facilidad_anuncio:40,
        escalabilidad:50,
        recencia_lanzamiento:30
      }
    };
  }

  function clamp(v){
    if(v<0) return 0; if(v>1) return 1; return v;
  }

  function normalizeMetric(name, value, ranges={}){
    if(value==null) return null;
    switch(name){
      case 'magnitud_deseo':
      case 'nivel_consciencia_headroom':
      case 'competition_level_invertido':
      case 'facilidad_anuncio':
      case 'escalabilidad':
      case 'durabilidad_recurrencia':
        return MAPS[name][String(value).toLowerCase()] ?? null;
      case 'evidencia_demanda':{
        const v = Math.log1p(Number(value)||0);
        const r = ranges[name]||{}; const min=r.p5??0; const max=r.p95??1; return clamp((v-min)/(max-min||1));
      }
      case 'tasa_conversion':
        return clamp((Number(value)||0)/100);
      case 'ventas_por_dia':{
        const v = Number(value)||0; const r=ranges[name]||{}; const min=r.p5??0; const max=r.p95??1; return clamp((v-min)/(max-min||1));
      }
      case 'recencia_lanzamiento':
        return Math.exp(-((Number(value)||0)/180));
      default:
        return null;
    }
  }

  function computeRanges(list){
    const nums={evidencia_demanda:[], ventas_por_dia:[]};
    list.forEach(p=>{
      if(p.evidencia_demanda!=null) nums.evidencia_demanda.push(Math.log1p(Number(p.evidencia_demanda)||0));
      if(p.ventas_por_dia!=null) nums.ventas_por_dia.push(Number(p.ventas_por_dia)||0);
    });
    const ranges={};
    for(const k in nums){
      const arr=nums[k].sort((a,b)=>a-b);
      if(arr.length){
        const p5=arr[Math.floor(arr.length*0.05)];
        const p95=arr[Math.floor(arr.length*0.95)];
        ranges[k]={p5,p95};
      }
    }
    return ranges;
  }

  function scoreProduct(prod, weights, ranges){
    let totalW=0; let score=0;
    for(const k in weights){
      const w=weights[k];
      const norm=normalizeMetric(k, prod[k], ranges);
      if(norm!=null){
        totalW+=w;
        score+=w*norm;
      }
    }
    if(totalW<=0) return 0;
    return score/totalW;
  }

  global.winnerV2 = {metricDefs, MAPS, normalizeMetric, computeRanges, scoreProduct, getDefaultWeightsV2};
})(window);

File: product_research_app/services/winner_v2.py
"""Winner Score v2 calculation utilities."""
from __future__ import annotations
import math
from typing import Dict, Any, Iterable

# Mapping tables for categorical metrics
MAGNITUD_DESEO = {"low":0.33, "medium":0.66, "high":1.0}
NIVEL_CONSCIENCIA_HEADROOM = {"unaware":1.0, "problem":0.8, "solution":0.6, "product":0.4, "most":0.2}
COMPETITION_LEVEL_INVERTIDO = {"low":1.0, "medium":0.5, "high":0.0}
FACILIDAD = {"low":0.33, "med":0.66, "medium":0.66, "high":1.0}
ESCALABILIDAD = FACILIDAD
DURABILIDAD = {"consumible":1.0, "durable":0.0, "intermedio":0.5}

MAPS = {
    "magnitud_deseo": MAGNITUD_DESEO,
    "nivel_consciencia_headroom": NIVEL_CONSCIENCIA_HEADROOM,
    "competition_level_invertido": COMPETITION_LEVEL_INVERTIDO,
    "facilidad_anuncio": FACILIDAD,
    "escalabilidad": ESCALABILIDAD,
    "durabilidad_recurrencia": DURABILIDAD,
}

ALL_METRICS = [
    "magnitud_deseo",
    "nivel_consciencia_headroom",
    "evidencia_demanda",
    "tasa_conversion",
    "ventas_por_dia",
    "recencia_lanzamiento",
    "competition_level_invertido",
    "facilidad_anuncio",
    "escalabilidad",
    "durabilidad_recurrencia",
]

def clamp(v: float) -> float:
    return 0.0 if v < 0 else 1.0 if v > 1 else v

def _percentiles(values: Iterable[float]) -> Dict[str, float]:
    vals = sorted(v for v in values if v is not None)
    if not vals:
        return {"p5":0.0, "p95":1.0}
    def p(q: float) -> float:
        idx = int(q * (len(vals)-1))
        return vals[idx]
    return {"p5": p(0.05), "p95": p(0.95)}

def compute_ranges(products: Iterable[Dict[str, Any]]) -> Dict[str, Dict[str,float]]:
    ev = []
    vpd = []
    for p in products:
        if p.get("evidencia_demanda") is not None:
            ev.append(math.log1p(float(p.get("evidencia_demanda", 0))))
        if p.get("ventas_por_dia") is not None:
            vpd.append(float(p.get("ventas_por_dia", 0)))
    return {
        "evidencia_demanda": _percentiles(ev),
        "ventas_por_dia": _percentiles(vpd),
    }

def normalize_metric(name: str, value: Any, ranges: Dict[str, Dict[str,float]]) -> float | None:
    if value is None:
        return None
    if name in MAPS:
        return MAPS[name].get(str(value).lower())
    if name == "evidencia_demanda":
        v = math.log1p(float(value))
        r = ranges.get(name, {})
        return clamp((v - r.get("p5",0.0)) / (r.get("p95",1.0) - r.get("p5",0.0) or 1))
    if name == "tasa_conversion":
        v = float(value)
        if v > 1:
            v /= 100.0
        return clamp(v)
    if name == "ventas_por_dia":
        v = float(value)
        r = ranges.get(name, {})
        return clamp((v - r.get("p5",0.0)) / (r.get("p95",1.0) - r.get("p5",0.0) or 1))
    if name == "recencia_lanzamiento":
        return math.exp(-float(value)/180.0)
    return None

def score_product(
    prod: Dict[str, Any],
    weights: Dict[str, float],
    ranges: Dict[str, Dict[str, float]] | None = None,
    missing: list[str] | None = None,
    used: list[str] | None = None,
) -> float | None:
    if ranges is None:
        ranges = compute_ranges([prod])
    total_w = 0.0
    score = 0.0
    for k, w in weights.items():
        if w <= 0:
            continue
        val = normalize_metric(k, prod.get(k), ranges)
        if val is None or (isinstance(val, float) and math.isnan(val)):
            if missing is not None:
                missing.append(k)
            continue
        if used is not None:
            used.append(k)
        score += w * val
        total_w += w
    if total_w <= 0:
        return None
    return score / total_w
